# Student Performance

Jade Chen 22411268

### Description

The data includes two datasets about student performance. The data is taken from two Portuguese secondary schools and collected through school reports and questionnaires. One dataset is about “Mathematics” and the other is about “Portuguese Language”.

For our project, we are only using the dataset on “Mathematics”.

### Research Question
<span style="color:red"> (Predict) Can family background and study habits predict a students' final grade in Mathematics?</span>

1. Response Variable (Dependent Variable)
    - `G3` (Final Grade)

2. Explanatory Variables (Independent Variables)
    - Family Background: `Internet` (Internet Access at Home)
    - Study Habits: `StudyTime` (Weekly Study Time)
    
* Note: There are multiple variables that can be used for both family background and study habits, here, it has been narrowed down to one variable each.

### Importing Packages

In [1]:
library(tidyverse)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


### Read Data
Split by delimiter `;` to have one feature per column.

In [12]:
data <- read_csv('student-mat.csv') |>
    separate_wider_delim(cols = 'school;sex;age;address;famsize;Pstatus;Medu;Fedu;Mjob;Fjob;reason;guardian;traveltime;studytime;failures;schoolsup;famsup;paid;activities;nursery;higher;internet;romantic;famrel;freetime;goout;Dalc;Walc;health;absences;G1;G2;G3',
                         delim = ";",
                         names = c('School', 'Sex', 'Age', 'Address', 'FamilySize', 'PStatus', 'MEdu', 'FEdu', 'MJob', 'FJob', 'Reason', 'Guardian', 'TravelTime', 'StudyTime', 'Failures', 'SchoolSupport', 'FamSupport', 'Paid', 'Activities', 'Nursery', 'Higher', 'Internet', 'Romantic', 'FamRel', 'FreeTime', 'GoOut', 'DAlch', 'WAlch', 'Health', 'Absences', 'G1', 'G2', 'G3')
                        )
head(data)

Rows: 395 Columns: 1
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): school;sex;age;address;famsize;Pstatus;Medu;Fedu;Mjob;Fjob;reason;g...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


School Sex Age Address FamilySize PStatus MEdu FEdu MJob     FJob     ⋯
1 GP     F   18  U       GT3        A       4    4    at_home  teacher  ⋯
2 GP     F   17  U       GT3        T       1    1    at_home  other    ⋯
3 GP     F   15  U       LE3        T       1    1    at_home  other    ⋯
4 GP     F   15  U       GT3        T       4    2    health   services ⋯
5 GP     F   16  U       GT3        T       3    3    other    other    ⋯
6 GP     M   16  U       LE3        T       4    3    services other    ⋯
  FamRel FreeTime GoOut DAlch WAlch Health Absences G1 G2 G3
1 4      3        4     1     1     3      6        5  6  6 
2 5      3        3     1     1     3      4        5  5  6 
3 4      3        2     2     3     3      10       7  8  10
4 3      2        2     1     1     5      2        15 14 15
5 4      3        2     1     2     5      4        6  10 10
6 5      4        2     1     2     5      10       15 15 15

### Number of Observations
There are 395 unique observations. With the `unique()` function, possible duplicate values present in data frame were eliminated. 

In [3]:
nrow(unique(data))

[1] 395

### Data Wrangling
For ease of the following descriptions, the relevant features were converted to numeric/factor type according to the "Additional Variable Information" from the dataset source.

In [4]:
#Binary -> Factor Type
data$School <- as.factor(data$School)
data$Sex <- as.factor(data$Sex)
data$Address <- as.factor(data$Address)
data$FamilySize <- as.factor(data$FamilySize)
data$PStatus <- as.factor(data$PStatus)
data$SchoolSupport <- as.factor(data$SchoolSupport)
data$FamSupport <- as.factor(data$FamSupport)
data$Paid <- as.factor(data$Paid)
data$Activities <- as.factor(data$Activities)
data$Nursery <- as.factor(data$Nursery)
data$Higher <- as.factor(data$Higher)
data$Internet <- as.factor(data$Internet)
data$Romantic <- as.factor(data$Romantic)

#Nominal -> Factor Type
data$MJob <- as.factor(data$MJob)
data$FJob <- as.factor(data$FJob)
data$Reason <- as.factor(data$Reason)
data$Guardian <- as.factor(data$Guardian)

#Numeric -> Numeric Type
data$Age <- as.numeric(data$Age)
data$MEdu <- as.numeric(data$MEdu)
data$FEdu <- as.numeric(data$FEdu)
data$TravelTime <- as.numeric(data$TravelTime)
data$StudyTime <- as.numeric(data$StudyTime)
data$Failures <- as.numeric(data$Failures)
data$FamRel <- as.numeric(data$FamRel)
data$FreeTime <- as.numeric(data$FreeTime)
data$GoOut <- as.numeric(data$GoOut)
data$DAlch <- as.numeric(data$DAlch)
data$WAlch <- as.numeric(data$WAlch)
data$Health <- as.numeric(data$Health)
data$Absences <- as.numeric(data$Absences)
data$G1 <- as.numeric(data$G1)
data$G2 <- as.numeric(data$G2)
data$G3 <- as.numeric(data$G3)

### Feature Information
In total, there are 33 features. The following dataframe shows the name, the number of unique values, and the type of each feature.

In [5]:
num_unique <- lapply(data, function(x) length(unique(x)))
column_types <- lapply(data, function(x) class(x))

result_df <- data.frame(
    NumUnique = unlist(num_unique),
    DataType = unlist(column_types)
)
result_df

NumUnique DataType
School         2        factor  
Sex            2        factor  
Age            8        numeric 
Address        2        factor  
FamilySize     2        factor  
PStatus        2        factor  
MEdu           5        numeric 
FEdu           5        numeric 
MJob           5        factor  
FJob           5        factor  
Reason         4        factor  
Guardian       3        factor  
TravelTime     4        numeric 
StudyTime      4        numeric 
Failures       4        numeric 
SchoolSupport  2        factor  
FamSupport     2        factor  
Paid           2        factor  
Activities     2        factor  
Nursery        2        factor  
Higher         2        factor  
Internet       2        factor  
Romantic       2        factor  
FamRel         5        numeric 
FreeTime       5        numeric 
GoOut          5        numeric 
DAlch          5        numeric 
WAlch          5        numeric 
Health         5        numeric 
Absences      34        numeric 
G1            17        numeric 
G2            17        numeric 
G3            18        numeric

### Unique Values of Factor Features
Furthermore, we can see the possible unique values of each factor feature.

In [6]:
data_factor <- select_if(data, is.factor)
ulst <- lapply(data_factor, function(x) unique(as.character(x)))
ulst

$School
[1] "GP" "MS"

$Sex
[1] "F" "M"

$Address
[1] "U" "R"

$FamilySize
[1] "GT3" "LE3"

$PStatus
[1] "A" "T"

$MJob
[1] "at_home"  "health"   "other"    "services" "teacher" 

$FJob
[1] "teacher"  "other"    "services" "health"   "at_home" 

$Reason
[1] "course"     "other"      "home"       "reputation"

$Guardian
[1] "mother" "father" "other" 

$SchoolSupport
[1] "yes" "no" 

$FamSupport
[1] "no"  "yes"

$Paid
[1] "no"  "yes"

$Activities
[1] "no"  "yes"

$Nursery
[1] "yes" "no" 

$Higher
[1] "yes" "no" 

$Internet
[1] "no"  "yes"

$Romantic
[1] "no"  "yes"

### Range of Numeric Features
For numeric features, we can further see the range of each feature.

In [7]:
# Subset numeric columns
data_numeric <- select_if(data, is.numeric)
data.frame(Min = sapply(data_numeric,min),
           Max = sapply(data_numeric,max))

Min Max
Age        15  22 
MEdu        0   4 
FEdu        0   4 
TravelTime  1   4 
StudyTime   1   4 
Failures    0   3 
FamRel      1   5 
FreeTime    1   5 
GoOut       1   5 
DAlch       1   5 
WAlch       1   5 
Health      1   5 
Absences    0  75 
G1          3  19 
G2          0  19 
G3          0  20

### Feature Details
Below is the detailed description for possible values of each 33 features.

1. `School`: Student's School
    - GP - Gabriel Pereira
    - MS - Mousinho da Silveira
2. `Sex`: Student's Sex
    - F - female
    - M - male
3. `Age`: Student's Age
    - from 15 to 22
4. `Address`: Student's Home Address Type
    - U - urban
    - R - rural
5. `FamilySize`: Family Size
    - LE3 - less or equal to 3
    - GT3 - greater than 3
6. `Pstatus`: Parent's Cohabitation Status
    - T - living together
    - A - apart
7. `MEdu`: Mother's Education
    - 0 - none
    - 1 - primary education (4th grade)
    - 2 – 5th to 9th grade
    - 3 – secondary education
    - 4 – higher education
8. `FEdu`: Father's Education
    - 0 - none
    - 1 - primary education (4th grade)
    - 2 – 5th to 9th grade
    - 3 – secondary education
    - 4 – higher education
9. `MJob`: Mother's Job
    - teacher
    - health - health care related
    - services - civil services (e.g. administrative or police)
    - at_home
    - other
10. `FJob`: Father's Job
    - teacher
    - health - health care related
    - services - civil services (e.g. administrative or police)
    - at_home
    - other
11. `Reason`: Reason to Choose This School
    - home - close to home
    - reputation - school reputation
    - course - course preference
    - other
12. `Guardian`: Student's Guardian
    - mother
    - father
    - other
13. `TravelTime`: Home to School Travel Time
    - 1 - <15 min.
    - 2 - 15 to 30 min.
    - 3 - 30 min. to 1 hour
    - 4 - >1 hour
14. `StudyTime`: Weekly Study Time
    - 1 - <2 hours
    - 2 - 2 to 5 hours
    - 3 - 5 to 10 hours
    - 4 - >10 hours
15. `Failures`: Number of Past Class Failures
    - n if: 1 <= n < 3
    - else: 4
16. `SchoolSup`: Extra Educational Support
    - yes
    - no
17. `FamSup`: Family Educational Support
    - yes
    - no
18. `Paid`: Extra Paid Classes within the Course Subject
    - yes
    - no
19. `Activities`: Extra-Curricular Activities
    - yes
    - no
20. `Nursery`: Attended Nursery School
    - yes
    - no
21. `Higher`: Wants to Take Higher Education
    - yes
    - no
22. `Internet`: Internet Access at Home
    - yes
    - no
23. `Romantic`: With a Romantic Relationship
    - yes
    - no
24. `FamRel`: Quality of Family Relationships
    - 1 (very bad) to 5 (excellent)
25. `FreeTime`: Free Time After School
    - 1 (very low) to 5 (very high)
26. `GoOut`: Going Out with Friends
    - 1 (very low) to 5 (very high)
27. `DAlch`: Workday Alchohol Consumption
    - 1 (very low) to 5 (very high)
28. `WAlch`: Weekend Alchohol Consumption
    - 1 (very low) to 5 (very high)
29. `Health`: Current Health Status
    - 1 (very bad) to 5 (very good)
30. `Absences`: Number of School Absences
    - 0 to 93
31. `G1`: First Period Grade
    - 0 to 20
32. `G2` Second Period Grade
    - 0 to 20
33. `G3` Final Grade
    - 0 to 20

In [8]:
# glimpse(data)

In [9]:
column_names <- colnames(data)